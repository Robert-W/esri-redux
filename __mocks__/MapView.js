module.exports = function MapView() {
  return Promise.resolve({});
};
