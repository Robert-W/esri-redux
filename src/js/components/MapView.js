import { viewCreated, getItemInfo, } from 'js/actions/mapActions';
import { MAP_OPTIONS, VIEW_OPTIONS, URLS, OUT_FIELDS } from 'js/config';
import LocateModal from 'js/components/modals/Locate';
import ShareModal from 'js/components/modals/Share';
import Spinner from 'js/components/shared/Spinner';
import Controls from 'js/components/Controls';
import MapView from 'esri/views/MapView';
import React, { Component } from 'react';
import appStore from 'js/appStore';
import FeatureLayer from 'esri/layers/FeatureLayer';
import {earthquakePopupTemplate, drinkingWaterPopupTemplate} from '../constants/popupTemplates'; //Import Popup Templates
import { getMap, getFeatureLayer } from '../utils/map-helper';
import {getTimestamp} from '../utils/timestamp-helper';

export default class Map extends Component {
  constructor(props) {
    super(props);
    this._createMapView = this._createMapView.bind(this);
    /*
      Create feature layers.
      Import templates for pop-ups.
      Specify outfields that we want to pull in.
    */
    this.earthquakeFeatureLayer = getFeatureLayer(URLS.featureLayer, OUT_FIELDS.earthquakeLayer, earthquakePopupTemplate);
    this.waterFeatureLayer = getFeatureLayer(URLS.waterFeatureLayer, OUT_FIELDS.waterLayer, drinkingWaterPopupTemplate);

}
  displayName: 'Map';
  state = appStore.getState();
  view = {};

  //Abstracted Create Map View function from lifecycle methods into a helper function
  _createMapView(map) {
    const promise = new MapView({
      container: this.refs.mapView,
      map: map,
      ...VIEW_OPTIONS
    });
    promise.then(view => {
      this.view = view;
      appStore.dispatch(viewCreated());
      //- Webmap from https://developers.arcgis.com/javascript/latest/api-reference/esri-WebMap.html
      //appStore.dispatch(getItemInfo('e691172598f04ea8881cd2a4adaa45ba'));
    });
  }

  //When the component unmounts, set the filters to null;
  componentWillUnmount() {
    this.earthquakeFeatureLayer.filter = null;
    this.waterFeatureLayer.filter = null;
    this.unsubscribe();

  }


  componentDidMount() {
    /*
      When the component mounts for the first time, make sure
      filters are null.
    */
    this.earthquakeFeatureLayer.filter = null;
    this.waterFeatureLayer.filter = null;
    // Subscribe to the store for updates
    this.unsubscribe = appStore.subscribe(this.storeDidUpdate);
    //Call helper function to get initial map and create view on mount:
    let map = getMap([this.earthquakeFeatureLayer, this.waterFeatureLayer], MAP_OPTIONS);
    this._createMapView(map);
  }

  /*
    Check to see if the filter props have changed when new props are received.
    If they have changed, apply (or remove) the filter from the appropriate feature
    layer.
  */
  UNSAFE_componentWillReceiveProps(nextProps) {
    /*
      If the filter is true, update the feature layer and call new
      render of map
    */
    if (nextProps.earthquakeFilter !== this.props.earthquakeFilter) {
      if (nextProps.earthquakeFilter) {
        const targetTime = getTimestamp();
        //this.earthquakeFeatureLayer.setDefinitionExpression("TIME=0");
        this.earthquakeFeatureLayer.filter = {
          where: `Time > ${targetTime}`
        };
        let map = getMap([this.earthquakeFeatureLayer, this.waterFeatureLayer], MAP_OPTIONS);
            this._createMapView(map);
      } else {
        this.earthquakeFeatureLayer.filter = null;
        let map = getMap([this.earthquakeFeatureLayer, this.waterFeatureLayer], MAP_OPTIONS);
        this._createMapView(map);
      }
    }
     /*
      If the filter is true, update the feature layer and call new
      render of map
    */
    if (nextProps.waterFilter !== this.props.waterFilter) {
      console.log('true, new water filter props', this.props.waterFilter);
      if (nextProps.waterFilter) {
        this.waterFeatureLayer.filter = {
              where: "amenity = 'drinking_water'"
            };
            //this.waterFeatureLayer.setDefinitionExpression("amenity='drinking_water'");
            let map = getMap([this.earthquakeFeatureLayer, this.waterFeatureLayer], MAP_OPTIONS);
            this._createMapView(map);

      } else {
        this.waterFeatureLayer.filter = null;
      }
    }
  }

  storeDidUpdate = () => {
    this.setState(appStore.getState());
  };

  render () {
    const {shareModalVisible, locateModalVisible} = this.state;

    return (
      <div ref='mapView' className='map-view'>
        <Controls view={this.view} />
        <Spinner active={!this.view.ready} />
        <ShareModal visible={shareModalVisible} />
        <LocateModal visible={locateModalVisible} />
      </div>
    );
  }
}
