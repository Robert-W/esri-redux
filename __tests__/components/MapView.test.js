import renderer from 'react-test-renderer';
import MapView from 'js/components/MapView';
import React from 'react';

/**
* Snapshot testing can be very useful if used correctly, this example just shows how to use
* it but this is not the most useful snapshot, sometimes it is better to call it multiple
* times inside a test and take snapshots of the component changing throughout some interactions,
* Ex. snapshot default state, snapshot after click/mouseover, snapshot after mouseleave
*/
describe('MapView snapshot Tests', () => {

  test('Sample of snapshot testing, not very useful, but shows how to do it', () => {
    const header = renderer.create(
      <MapView title="Snapshot Test #1" subtitle="Test output over time in your component" />
    );
    const tree = header.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
