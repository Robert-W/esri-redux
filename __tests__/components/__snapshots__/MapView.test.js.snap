// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView snapshot Tests Sample of snapshot testing, not very useful, but shows how to do it 1`] = `
<div
  className="map-view"
>
  <div
    className="map-controls shadow"
  >
    <ul
      className="map-controls__list"
    >
      <li
        className="map-controls__item pointer"
        onClick={[Function]}
      >
        <svg
          aria-label="Zoom out"
          className="map-controls__item-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<use xlink:href=\\"#icon-zoom-out\\" />",
            }
          }
          role="img"
        />
      </li>
      <li
        className="map-controls__item pointer"
        onClick={[Function]}
      >
        <svg
          aria-label="Zoom in"
          className="map-controls__item-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<use xlink:href=\\"#icon-zoom-in\\" />",
            }
          }
          role="img"
        />
      </li>
      <li
        className="map-controls__item pointer"
        onClick={[Function]}
      >
        <svg
          aria-label="Share your experience"
          className="map-controls__item-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<use xlink:href=\\"#icon-share\\" />",
            }
          }
          role="img"
        />
      </li>
      <li
        className="map-controls__item pointer"
        onClick={[Function]}
      >
        <svg
          aria-label="Find my location"
          className="map-controls__item-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<use xlink:href=\\"#icon-locate\\" />",
            }
          }
          role="img"
        />
      </li>
    </ul>
  </div>
  <div
    style={
      Object {
        "display": "block",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": "50px",
          "left": "50%",
          "position": "absolute",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "width": "50px",
        }
      }
    >
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "
                        <g transform='translate(25,25) rotate(-90)'>
                          <path d=\\"M0,25A25,25 0 1,1 0,-25A25,25 0 1,1 0,25M0,20A20,20 0 1,0 0,-20A20,20 0 1,0 0,20Z\\" style=\\"fill:transparent\\"></path>
                          <path
                            d=\\"M1.5308084989341915e-15,-25A25,25 0 0,1 25,0L20,0A20,20 0 0,0 1.2246467991473533e-15,-20Z\\"
                            transform=\\"rotate(709.287459262793)\\"
                            style=\\"animation: spinner 0.95s cubic-bezier(0.645, 0.045, 0.355, 1.000) infinite;fill: #FFFFFF;\\">
                          </path>
                        </g>
                      ",
          }
        }
        height="50"
        width="50"
      />
    </div>
  </div>
  <div
    style={
      Object {
        "display": "none",
        "height": "100%",
        "left": 0,
        "position": "fixed",
        "top": 0,
        "width": "100%",
        "zIndex": 10,
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
          "height": "100%",
          "width": "100%",
        }
      }
    />
    <article
      className="share-modal"
      style={
        Object {
          "backgroundColor": "white",
          "height": "auto",
          "left": "50%",
          "padding": "30px",
          "position": "absolute",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        }
      }
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
            "display": "flex",
            "height": "30px",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": "30px",
          }
        }
        title="close"
      >
        <svg
          style={
            Object {
              "fill": "#555555",
              "height": "26px",
              "margin": "auto",
              "width": "26px",
            }
          }
          viewBox="0 0 25 25"
        >
          <title>
            Close
          </title>
          <path
            d="M 5 19 L 19 5 L 21 7 L 7 21 L 5 19 ZM 7 5 L 21 19 L 19 21 L 5 7 L 7 5 Z"
          />
        </svg>
      </div>
      <div
        style={
          Object {
            "maxHeight": "450px",
            "overflowX": "hidden",
            "overflowY": "auto",
            "padding": "5px",
          }
        }
      >
        <h3>
          Share Something
        </h3>
      </div>
    </article>
  </div>
  <div
    style={
      Object {
        "display": "none",
        "height": "100%",
        "left": 0,
        "position": "fixed",
        "top": 0,
        "width": "100%",
        "zIndex": 10,
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
          "height": "100%",
          "width": "100%",
        }
      }
    />
    <article
      className="locate-modal"
      style={
        Object {
          "backgroundColor": "white",
          "height": "auto",
          "left": "50%",
          "padding": "30px",
          "position": "absolute",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        }
      }
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
            "display": "flex",
            "height": "30px",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": "30px",
          }
        }
        title="close"
      >
        <svg
          style={
            Object {
              "fill": "#555555",
              "height": "26px",
              "margin": "auto",
              "width": "26px",
            }
          }
          viewBox="0 0 25 25"
        >
          <title>
            Close
          </title>
          <path
            d="M 5 19 L 19 5 L 21 7 L 7 21 L 5 19 ZM 7 5 L 21 19 L 19 21 L 5 7 L 7 5 Z"
          />
        </svg>
      </div>
      <div
        style={
          Object {
            "maxHeight": "450px",
            "overflowX": "hidden",
            "overflowY": "auto",
            "padding": "5px",
          }
        }
      >
        <h3>
          Locate Something
        </h3>
      </div>
    </article>
  </div>
</div>
`;
